Add new `tctl bots update` command

timothyb89 opened the pull request:
This adds a new command to update roles and logins for existing bots, with flags to both set and append as desired.

changelog: Add new `tctl bots update` command to update bot roles and logins

Closes #34262

timothyb89 committed:
Add new `tctl bots update` command

This adds a new command to update roles and logins for existing bots,
with flags to both set and append as desired.

changelog: Add new `tctl bots update` command to update bot roles and logins

timothyb89 committed:
Pluralize "add-" flags for consistency

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tctl-update-bot

timothyb89 committed:
Add docs for `tctl bots update`

Also merges in some other misc docs changes for `tctl bots ...`

timothyb89 committed:
Fix docs lint

timothyb89 committed:
Fix incorrect example

strideynet commented:
Should we be using the new UpdateBot RPC for this instead ? Or is this so we can backport this further?

timothyb89 committed:
Apply suggestions from code review

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

timothyb89 committed:
Use the bot service's `UpdateBot()` RPC instead of modifying directly

This switches to use the new `UpdateBot()` RPC instead of modifying
the bot user and role directly. Also, tweaks CLI messaging to be a
bit less chatty.

timothyb89 commented:
> Should we be using the new UpdateBot RPC for this instead ? Or is this so we can backport this further?

Ah, good call, truthfully I forgot we had an explicit update RPC now :sweat_smile:.

Anyway, I've updated it to use the correct RPC.

hugoShaka commented on tool/tctl/common/bots_command.go:
```diff
@@ -483,7 +497,142 @@ func (c *BotsCommand) LockBot(ctx context.Context, client auth.ClientI) error {
+		return trace.Wrap(err)
+	}
+
+	log.Infof("Bot %q has been updated. Roles will take affect on its next renewal.", c.botName)
```
```suggestion
log.Infof("Bot %q has been updated. Roles will take effect on its next renewal.", c.botName)
```

hugoShaka commented on tool/tctl/common/bots_command.go:
```diff
@@ -494,3 +643,52 @@ func splitRoles(flag string) []string {
 	return roles
 }
+
+// setsEqual determines if two sets contain the same keys.
+func setsEqual[T comparable](a map[T]struct{}, b map[T]struct{}) bool {
```
This set implementation is pretty clean and generic, can we move it to some shared lib? I'm constantly reimplementing sets with empty maps and would benefit from this code.

timothyb89 replied:
I've spun it out into `github.com/gravitational/teleport/lib/utils/set` to make it more reusable (and adjusted the API to make it a bit more sane than flinging `map[T]struct{}{}` around)

timothyb89 committed:
Update tool/tctl/common/bots_command.go

Co-authored-by: Hugo Shaka <hugo.hervieux@goteleport.com>

timothyb89 committed:
Move set functions into their own reusable package

The set package adds a little extra functionality but is still
missing various important functions, like subset/superset,
difference, intersection, xor/symmetric difference, etc.

This changes the API slightly, so the tctl command is updated to use
the new package.

timothyb89 committed:
Merge branch 'timothyb89/tctl-update-bot' of github.com:gravitational/teleport into timothyb89/tctl-update-bot

timothyb89 commented on tool/tctl/common/bots_command.go:
```diff
@@ -494,3 +643,52 @@ func splitRoles(flag string) []string {
+	}
+
+	return true
+}
```
A bit, but not meaningfully. I would hope the compiler will optimize the useless value checks away, but even if not they probably aren't costly vs having a duplicate implementation. I've switched to it in the new library-ified `set` package.

timothyb89 commented on tool/tctl/common/bots_command.go:
```diff
@@ -494,3 +643,52 @@ func splitRoles(flag string) []string {
+		ret = append(ret, entry)
+	}
+
+	return ret
```
Good call, done.

timothyb89 committed:
Remove duplicate tests

timothyb89 committed:
Reword flag help strings for clarity

timothyb89 commented on tool/tctl/common/bots_command.go:
```diff
@@ -88,6 +93,13 @@ func (c *BotsCommand) Initialize(app *kingpin.Application, config *servicecfg.Co
+	c.botsUpdate.Arg("name", "Name of an existing bot to update.").Required().StringVar(&c.botName)
+	c.botsUpdate.Flag("set-roles", "A comma-separated list of roles to replace.").StringVar(&c.botRoles)
+	c.botsUpdate.Flag("add-roles", "A comma-separated list of roles to add to an existing bot.").StringVar(&c.addRoles)
+	c.botsUpdate.Flag("set-logins", "A comma-separated list of allowed logins to replace").StringVar(&c.setLogins)
```
I've reworded the help strings to be a bit explicit about what they're meant to do.

codingllama commented on lib/utils/set/set.go:
```diff
@@ -0,0 +1,76 @@
+import "golang.org/x/exp/maps"
+
+// Set represents a set data structure, using the map type with an empty value.
+type Set[T comparable] map[T]struct{}
```
Hey folks, let's not a push a custom Set structure.

Sets are typically implemented as a valueless map, so in Go we just use a plain `map[Foo]struct{}` when that is necessary (much in the way this is implemented).

Most of the added methods are available as functions in `maps` and others, like Union, can be "implemented" as a `for` if/when needed.

What I'm trying to say is that the language "lacks" Set on purpose, as `map` does its job. Let's not muddle it by adding a custom type to the mix.

timothyb89 replied:
I guess I'd just like to interrogate this a bit, because there's clearly some interest in not having to rewrite this same code all the time. This isn't even the first halfway implementation I've personally written in Teleport and in any other non-datastructure context I'd appropriately receive review feedback requesting that I didn't duplicate code.

Objectively speaking, most set functionality is _not_ available in `maps`. It provides (with caveats) `Equals` and `Keys` and... nothing else. No set methods are hugely complex in isolation but it is a real drag to keep reimplementing them.

Given that, which of these if any would be more appropriate:
1. A collection of generic set utility functions that operate directly on `map[T]struct{}` in a `sets` package, for example `sets.FromArray`, `sets.ToArray`, `sets.Union`, `sets.Intersection`, `sets.Difference`, `set.IsSubset`, etc. Notably this does not create a new `type Set[...]`.
2. Use of a well-documented and well-tested third-party library providing set functionality, e.g. https://github.com/hashicorp/go-set
3. None of the above. We should always rewrite and retest all set functionality for each use-case.

codingllama replied:
I rarely find that the "entire functionality" of set is necessary. Taking a snippet from the current PR, with changed lines in green:

```diff
+	var currentLogins map[string]struct{}
+	if logins, ok := traits[constants.TraitLogins]; ok {
+		for _, login := range logins {
+			currentLogins[login] = struct{}
+		}
+	}

+	var desiredLogins map[string]struct{}
+	if c.setLogins != "" {
+		for _, login := range splitEntries(c.setLogins) {
+			desiredLogins[login] = struct{}
+		}
+	}

	addLogins := splitEntries(c.addLogins)
+	for _, login := range addLogins {
+		desiredLogins[login] = struct{}
+	}

+	if maps.Equal(currentLogins, desiredLogins) {
		log.Infof("Requested logins match existing, nothing to do: %+v", desiredLogins.ToArray())
		return nil
	}
```

I don't find that much different, or that much more code. We could consider creating a few utilities, or even importing an external library, but I would look for significant need before those either of those.

If you find yourself forced into constant transformations, it's also worth considering whether a different RPC or entity design would mitigate those problems as well.

zmb3 replied:
I tend to agree with Alan but I don't have strong feelings.

Would be nice to avoid x/exp/maps though. (We already use depguard to prevent against using x/exp/slices)

timothyb89 replied:
I... am not sure that I'm totally convinced, but alright, I've unrolled everything and removed all the helper functions.

> Would be nice to avoid x/exp/maps though. (We already use depguard to prevent against using x/exp/slices)

Yep, and as Hugo pointed out, the one function we wanted from it (`Keys()`) is being removed anyway. I've just unrolled that function too.

codingllama replied:
Thanks, Tim. Let me know if this causes frustration again, I'm interested to see the scenarios where it happens. Also let me know if you think I could help.

> the one function we wanted from it (Keys()) is being removed anyway

I think the usual answer is to iterate, like you did.

FWIW I think it's alright to pull "Keys" and "Values" into our codebase somewhere (I would do it in another PR though). I wouldn't be too surprised to see they land on `maps` sometime either, although I'm not sure what the official stance in regards to those is.

timothyb89 commented on lib/utils/set/set.go:
```diff
@@ -0,0 +1,76 @@
+
+package set
+
+import "golang.org/x/exp/maps"
```
Unfortunately the base `maps` package does not contain `Keys()`

hugoShaka replied:
`Keys()` will be dropped in 1.21 https://github.com/golang/go/issues/61538 :/

timothyb89 replied:
The `exp` import is removed now.

timothyb89 committed:
Remove set abstraction

Removes the Set package and unrolls all related function calls in
bots_command.

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tctl-update-bot

jimbishopp commented on tool/tctl/common/bots_command.go:
```diff
@@ -483,7 +498,165 @@ func (c *BotsCommand) LockBot(ctx context.Context, client auth.ClientI) error {
-func splitRoles(flag string) []string {
+// updateBotLogins applies updates from CLI arguments to a bot's logins trait,
+// updating the field mask if any updates were made.
+func (c *BotsCommand) updateBotLogins(ctx context.Context, bot *machineidv1pb.Bot, mask *fieldmaskpb.FieldMask) error {
```
It would be great to add tests for both updateBotRoles and updateBotLogins. The logins func should be straightforward. The roles func could be easy if the code that created desired roles was isolated in its own func (so you don’t have to include auth.ClientI in the test). updateBotRoles could call the new func to get desired roles then call the client. 

The other thing to consider is that this code may be simplified by not using maps at all. Both funcs essentially follow a similar pattern. Compare the current unique list of things to a new unique list of things and set the new list if they are different. Since all the things (logins and roles) are already lists and we need to create a new list, sticking with a sorted slice has the potential to reduce complexity. We’re not dealing with a huge amount of items in each list so performance of a sorting algorithm should be acceptable. 

Here’s an untested and naive example for the logins func. LMK what you think.

```go
func (c *BotsCommand) updateBotLogins(ctx context.Context, bot *machineidv1pb.Bot, mask *fieldmaskpb.FieldMask) error {
	traits := bot.Spec.GetTraits()

	index := -1
	var current []string
	for i, t := range traits { // []*Trait
		if t.Name == constants.TraitLogins {
			index = i
			current = t.Values
			break
		}
	}
	sort.Strings(current)

	desired := uniqStrings(splitEntries(c.setLogins)) // sorted []string
	if len(desired) == 0 {
		desired = slices.Clone(current)
	}
	desired = uniqStrings(append(desired, splitEntries(c.addLogins)...))

	if slices.Equal(current, desired) {
		return nil
	}

	if len(desired) > 0 {
		traits[index].Values = desired
	} else if len(traits) > 0 && index != -1 {
		traits = append(traits[:index], traits[index+1:]...)
	}
	bot.Spec.Traits = traits

	return trace.Wrap(mask.Append(&machineidv1pb.Bot{}, "spec.traits"))
}

func uniqStrings(list []string) []string {
	sort.Strings(list)

	var uniq []string
	for i := range list {
		if i == 0 || list[i] != list[i-1] {
			uniq = append(uniq, list[i])
		}
	}
	return uniq
}
```

timothyb89 replied:
> It would be great to add tests for both updateBotRoles and updateBotLogins.

Sure, I've added tests for both functions. It was pretty straightforward to test them in situ with a minimal client interface, so the structure is mostly left the same.

> The other thing to consider is that this code may be simplified by not using maps at all.

I do agree this approach is reasonable, but I think there is precedent for using maps here to deduplicate and compare the lists, and would probably consider it the idiomatic way of doing so in Go. We have [hundreds](https://github.com/search?q=repo%3Agravitational%2Fteleport+%2Fmap%5C%5Bstring%5C%5Dstruct%5C%7B%5C%7D%2F&type=code) of string sets throughout Teleport, many of which are used to deduplicate and compare string lists (and including several specialized `set` implementations).

That said, while searching I did discover the preexisting `utils.StringsSliceFromSet()` which reduces LoC a bit, and sorts the log output, which is nice for end users.

jimbishopp reviewed (commented):
Hey Tim! A couple of thoughts on this. Sorry for the last minute quarterbacking. :)

timothyb89 committed:
Add unit tests for `updateBotLogins()` and `updateBotRoles()`

This also makes a minimal API wrapper interface for mocking.

timothyb89 committed:
Use utils function to convert sets to a string slice

zmb3 commented on docs/pages/reference/cli/tctl.mdx:
````diff
@@ -441,6 +450,49 @@ $ tctl bots rm <name> [<flags>]
+Replace the bot `example` roles and add a new allowed Unix login:
+
+```code
+$ tctl bots update example --add-logins root --set-roles=access
````
```suggestion
$ tctl bots update example --add-logins=root --set-roles=access
```

Be consistent. Either use `--foo=bar` or `--foo bar`, but don't mix both in the same command :-)

timothyb89 replied:
Ah, good catch!

zmb3 reviewed (approved):
Just one docs nit.

timothyb89 committed:
Update docs/pages/reference/cli/tctl.mdx

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>
